src/content/symbol-store-db.js

```diff
@@ -1,134 +1,250 @@
+              reject(updateDateReq.error);
+            };
+          } else {
+            reject(new Error('This library does not exist .'));
+          }
+        };
+      });
+    });
   }
 
-  _setupDB(dbName) {
+  _setupDB(dbName: string): Promise<IDBDatabase> {
     return new Promise((resolve, reject) => {
-      const openreq = indexedDB.open(dbName, 1);
+      const openreq = window.indexedDB.open(dbName, 5);
       openreq.onerror = reject;
       openreq.onupgradeneeded = () => {
         const db = openreq.result;
         db.onerror = reject;
-        const tableStore = db.createObjectStore('symbol-tables', { autoIncrement: true });
-        tableStore.createIndex('libKey', ['debugName', 'breakpadId'], { unique: true });
+        db.deleteObjectStore('symbol-tables');
+        const store = db.createObjectStore('symbol-tables', { autoIncrement: true });
+        store.createIndex('libKey', ['debugName', 'breakpadId'], { unique: true });
+        store.createIndex('lastUsedDate', 'lastUsedDate', { unique: false });
       };
       openreq.onsuccess = () => {
-        this._db = openreq.result;
-        resolve();
+        const db = openreq.result;
+        resolve(db);
+        this._deleteAllBeforeDate(db, new Date(+new Date - this._maxAge));
       };
     });
   }
 
-  getLibKey(debugName, breakpadId) {
-    if (!this._db) {
-      return this._setupDBPromise.then(() => this.getLibKey(debugName, breakpadId));
-    }
+  // The utility functions below use both callback functions and promises.
+  // That's because IndexedDB transactions auto-close at the end of the
+  // current event tick if there hasn't been a new request after the last
+  // success event. So we need to synchronously add more work inside the
+  // onsuccess handler, and we do that by calling the callback function.
+  // Resolving a promise only calls any then() callback at the next MicroTick,
+  // and by that time the transaction will already have closed.
+  // We use the returned promise to asynchronously propagate errors.
+  _deleteRecords(store: IDBObjectStore, primaryKeys: number[], callback: () => void): Promise<void> {
+    return new Promise((resolve, reject) => {
+      const leftToDelete = primaryKeys.slice(0);
+      function deleteNext() {
+        if (leftToDelete.length === 0) {
+          callback();
+          resolve();
+          return;
+        }
+        const primaryKey = leftToDelete.shift();
+        const deleteReq = store.delete(primaryKey);
+        deleteReq.onsuccess = () => {
+          deleteNext();
+        };
+        deleteReq.onerror = () => {
+          reject(deleteReq.error);
+        };
+      }
+      deleteNext();
+    });
+  }
 
+  _deleteAllBeforeDate(db: IDBDatabase, beforeDate: Date): Promise<void> {
     return new Promise((resolve, reject) => {
-      const transaction = this._db.transaction('symbol-tables', 'readonly');
+      const transaction = db.transaction('symbol-tables', 'readwrite');
```

Comment:
That's a very good point. I hadn't really realized that one failing request will make the whole transaction error out. Having error handlers on the individual requests allows us to do more precise reporting of what caused the error, but we're not doing that anyway.